### PR TITLE
feat: fast path support on set operations within usersets

### PR DESCRIPTION
**Pull request by elbuo8**
## Description

Userset version/support of https://github.com/openfga/openfga/pull/2075

## References
https://github.com/openfga/openfga/issues/1931

## Review Checklist
- [ ] I have clicked on ["allow edits by maintainers"](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [ ] I have added documentation for new/changed functionality in this PR or in a PR to [openfga.dev](https://github.com/openfga/openfga.dev) [Provide a link to any relevant PRs in the references section above]
- [ ] The correct base branch is being used, if not `main`
- [ ] I have added tests to validate that the change in functionality is working as expected

If you haven't done so yet, we would appreciate it if you could star the [OpenFGA repository](https://github.com/openfga/openfga). :)

### CODE REVIEW

File: pkg/typesystem/typesystem_test.go [hunk 1]
```diff
@@ -3832,6 +3832,368 @@ func TestUsersetCanFastPath(t *testing.T) {
...
+				  schema 1.1
+				type user
+                type employee
+				type group
+				  relations
+                    define testers: [employee]
+					define assignee: [user, group#testers]
+			    type folder
+				  relations
+				    define allowed: [group#assignee]`,
```
Comment (miparnisari) [inline pkg/typesystem/typesystem_test.go:4149]: here and elsewhere, please fix the alignment

File: pkg/typesystem/typesystem_test.go [hunk 2]
```diff
@@ -3832,6 +3832,368 @@ func TestUsersetCanFastPath(t *testing.T) {
...
+						type folder
+							relations
+								define allowed: [group#member, group#owner]`,
+			objectType: "folder",
+			relation:   "allowed",
+			userType:   "user",
+			expected:   true,
+		},
+		{
+			name: "multiple_relation_references_some_complex",
```
Comment (miparnisari) [inline pkg/typesystem/typesystem_test.go:4004]: this test doesn't involve usersets like the others 🤔

File: pkg/typesystem/typesystem.go [hunk 3]
```diff
@@ -568,6 +569,78 @@ func (t *TypeSystem) IsDirectlyRelated(target *openfgav1.RelationReference, sour
...
+	// find all userset edges with valid weight
+	// but exit immediately if there is any above weight
+	for len(usersetEdges) != totalAllowed && len(edges) != 0 {
+		innerEdges := make([]*graph.WeightedAuthorizationModelEdge, 0)
+		for _, edge := range edges {
+			// edge is a set operator thus we have to inspect each node of the operator
+			if edge.GetEdgeType() == graph.RewriteEdge {
+				operationalEdges, ok := t.authzWeightedGraph.GetEdgesByNode(edge.GetTo())
+				if !ok {
+					fmt.Println("no edges 2")
```
Comment (jpadilla) [inline pkg/typesystem/typesystem.go:618]:
```suggestion
```

File: internal/graph/check_fast_path.go [hunk 4]
```diff
@@ -616,6 +616,47 @@ func (c *LocalChecker) resolveFastPath(ctx context.Context, leftChans []chan *it
...
+	defer span.End()
+
+	typesys, _ := typesystem.TypesystemFromContext(ctx)
+	objectType := tuple.GetType(req.GetTupleKey().GetObject())
+
+	cancellableCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	directlyRelatedUsersetTypes, _ := typesys.DirectlyRelatedUsersets(objectType, req.GetTupleKey().GetRelation())
+
+	leftChans := make([]chan *iteratorMsg, 0)
```
Comment (miparnisari) [inline internal/graph/check_fast_path.go:630]: nit: 

```suggestion
	leftChans := make([]chan *iteratorMsg, 0, len(directlyRelatedUsersetTypes))
```

File: internal/graph/check_fast_path.go [hunk 5]
```diff
@@ -616,6 +616,47 @@ func (c *LocalChecker) resolveFastPath(ctx context.Context, leftChans []chan *it
...
+			return nil, err
+		}
+		leftChans = append(leftChans, leftChan)
+	}
+
+	if len(leftChans) == 0 {
+		return &ResolveCheckResponse{
+			Allowed: false,
+		}, nil
+	}
```
Comment (miparnisari) [inline internal/graph/check_fast_path.go:655]: Could we write a test covering this block?

File: pkg/typesystem/typesystem.go [hunk 6]
```diff
@@ -568,6 +569,77 @@ func (t *TypeSystem) IsDirectlyRelated(target *openfgav1.RelationReference, sour
...
+				}
+				innerEdges = append(innerEdges, operationalEdges...)
+			}
+
+			// each edge must belong to one of the directly assignable userset types AND each one of them
+			// must not have a weight higher than the threshold/level. if true, collect as _all entries_ need to be accounted for
+			if edge.GetEdgeType() == graph.DirectEdge && allowed.Contains(edge.GetTo().GetUniqueLabel()) {
+				if len(edge.GetWildcards()) != 0 {
+					return false
+				}
```
Comment (miparnisari) [inline pkg/typesystem/typesystem.go:628]: could we add a test covering this block, one for `usersetEdges = append(usersetEdges, edge)`, and one for

```
if len(innerEdges) == 0 {
	break
}
```

?

File: internal/graph/check_fast_path.go (hunk 4 above)
Comment (elbuo8) [reply internal/graph/check_fast_path.go:630]: Can't. Had this in the original version of TTU but caused goroutine leak given that the consumer gets the events out of order. So the close can be received before processing all messages.

**Review by miparnisari — APPROVED** [review]

**Review by miparnisari — APPROVED** [review]

File: pkg/typesystem/typesystem.go (hunk 6 above)
Comment (miparnisari) [reply pkg/typesystem/typesystem.go:628]: https://github.com/openfga/openfga/issues/2119

File: internal/graph/check_fast_path.go (hunk 5 above)
Comment (miparnisari) [reply internal/graph/check_fast_path.go:655]: https://github.com/openfga/openfga/issues/2119
